pkg/image/admission/imagepolicy/imagepolicy.go

```diff
@@ -0,0 +1,392 @@
+	if !accept && !adjust {
+		return nil
+	}
+
+	podSpec, basePath, err := meta.GetPodSpec(attr.GetObject())
+	if err != nil {
+		return apierrs.NewForbidden(gr, attr.GetName(), fmt.Errorf("unable to apply image policy against objects of type %T, because they do not have a pod spec", attr.GetObject()))
+	}
+
+	covered := make(policyAttributes)
+
+	if accept {
+		if err := a.accept(podSpec, basePath, attr, covered); err != nil {
+			return err
+		}
+	}
+	if adjust {
+		if err := a.adjust(podSpec, attr, covered); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *imagePolicyPlugin) accept(podSpec *kapi.PodSpec, basePath *field.Path, attr admission.Attributes, covered policyAttributes) error {
+	gr := attr.GetResource().GroupResource()
+	requiresImage := a.accepter.RequiresImage(gr)
+	resolvesImage := a.accepter.ResolvesImage(gr)
+
+	var errs field.ErrorList
+	path := basePath.Child("containers")
+	for i := range podSpec.Containers {
+		if ok, err := a.acceptContainer(&podSpec.Containers[i], gr, covered, requiresImage, resolvesImage); !ok {
+			if err != nil {
+				errs = append(errs, field.Forbidden(path.Index(i).Child("image"), fmt.Sprintf("this image is prohibited by policy: %v", err.Error())))
+				continue
+			}
+			errs = append(errs, field.Forbidden(path.Index(i).Child("image"), "this image is prohibited by policy"))
+		}
+	}
+	path = basePath.Child("initContainers")
+	for i := range podSpec.InitContainers {
+		if ok, err := a.acceptContainer(&podSpec.InitContainers[i], gr, covered, requiresImage, resolvesImage); !ok {
+			if err != nil {
+				errs = append(errs, field.Forbidden(path.Index(i).Child("image"), fmt.Sprintf("this image is prohibited by policy: %v", err.Error())))
+				continue
+			}
+			errs = append(errs, field.Forbidden(path.Index(i).Child("image"), "this image is prohibited by policy"))
+		}
+	}
+	if len(errs) > 0 {
+		glog.V(5).Infof("failed to create: %v", errs)
+		return apierrs.NewInvalid(attr.GetKind().GroupKind(), attr.GetName(), errs)
+	}
+	glog.V(5).Infof("allowed: %#v", attr)
+	return nil
+}
+
+func (a *imagePolicyPlugin) acceptContainer(container *kapi.Container, gr unversioned.GroupResource, covered policyAttributes, requiresImage, resolvesImage bool) (bool, error) {
+	name := container.Image
+
+	// only resolve the image once per pod
+	if covered.Has(name) {
+		return true, nil
+	}
+	// explicitly allow empty strings to bypass checks
+	if len(name) == 0 {
+		return true, nil
+	}
+
+	ref, err := imageapi.ParseDockerImageReference(name)
+	if err != nil {
+		return false, err
+	}
+
+	// check the decision cache if we've already made a decision on the image
+	// TODO: time bound the decision cache so that namespace overrides can take effect
+	isDigest := len(ref.ID) > 0
+
+	var image *imageapi.Image
+	if requiresImage || (!isDigest && resolvesImage) {
+		image, err = a.resolver.Resolve(ref)
+		switch {
+		case apierrs.IsNotFound(err) && !resolvesImage:
```

Comment:
RequiresImage may be misnamed - requires is "the rules need us to try to load the image".  That's not the same as "the rules will fail if the image isn't loaded".  The rules make the decision about whether to fail on missing image.